dashboard/lib/build_dashboard_page.dart

```diff
@@ -218,14 +225,18 @@ class BuildDashboardPageState extends State<BuildDashboardPage> {
         ),
         onChanged: (String? selectedBranch) {
           branch = selectedBranch;
-          _updateNavigation(context);
+          _updateNavigation(context, _buildState);
         },
-        items: _buildState.branches.map<DropdownMenuItem<String>>((String value) {
+        items: _buildState.branches
+            .where((Branch b) => b.repository == _buildState.currentRepo)
```

Comment:
Is the `where` needed? The get-branches called should only return relevant data

Reply:
Yeah currently get-branches will return all branches with recent activities. Does it sound right if we add a url parameter 'repo', so that get-branches will filter based on repository during web request?

Reply:
That SGTM

Reply:
Our current logic is we run fetchBranches() once during setup, and retrieve all the branches we want. Later for frontend display, we just use `where` to select the branches we interested in from the pre-retrieved branches. So in total we only need one api request.

If we remove the `where` logic, we would have to send an api request whenever the user changes branch, and let updateBranch trigger fetchBranch(). Maybe i am missing something, I think it might be more efficient if we just pre-retrieve the branches and serve the user with what they need, instead of sending multiple api requests?

Reply:
You can update `updateCurrentRepoBranch` to make the call for `_fetchBranches`

Reply:
I guess one of the things removing `where` statements will also affect, is the logic to transit states. Currently, in order for dashboard to not break when user toggle the repo dropdown, (in case where user's current branch selection might be invalid for the new repo selection), I stored and carried over the latest branch selection for user, and combine with all the valid branch options that would come from the newest repository selection. This approach depends on a where statement to remove duplicate branches when combining the branch options.

If we remove the `where` logic, we would need to find another way for dashboard to not break. We can also list all branch options for all repos in the branch dropdown, but I think that would be less ideal.

Reply:
I wouldn't worry about this case. If users are manually overriding the branch, it's fine to reset it when the repo is swapped. Granted, I don't see why we can't keep the branch the same.

Reply:
umm previously I reset it to the default branch when switching repos, but that broke the dashboard from supporting invalid combination of repo+branch. To keep branch the same while swapping repositories, we need to support invalid combinations. 

And to add more context, I am trying to advocate for a `where` statement because of how dropdown button is built. The options in the dropdown button list at any time should always have an entry that matches current branch. Therefore if we want to keep the branch same after swapping repository, along with all the valid options brought by the new repository, we keep track of user's previous selection of branch (this selection is carried over from repo to repo), and combine user's previous selection of branch (which might be invalid in the new settings) with the currently valid list of options. And during this combination, we want to check if there are any duplicates in dropdownMenuItem, and I think we need a where statement to check that.

Reply:
I implemented this logic in the newest commit, maybe the code in  https://github.com/flutter/cocoon/pull/1695/commits/c7dbc53a13111e388a23c4ec212bca6ef2b51137#diff-795424fcac01b97b0d9ed4cff372972e6c6280b81e293264a85d3c8123106886R226-R251 would help explain why I think a `where` statement is necessary